### PR TITLE
#167190529 Validate User Signup Inputs

### PR DESCRIPTION
**Pull request by abdulfatai360**
#### What does this PR do?
- Write schema for user signup inputs
- Validate the user signup inputs against the schema

#### Description of Task proposed in this pull request?
- Create the file `schema/userSignup.js` inside the `/server` directory
- Install `@hapi/joi` library to write the schema and validate the signup inputs
- Install `sinon` to help in mocking `ExpressJs` `req`, `res`, and `next` objects for unit testing the validation middleware
- Check user signup inputs for validation errors
- Return validation errors to users, if any

#### How should this be manually tested (Quality Assurance)?
- Run git clone `https://github.com/andela/ah-rambo-backend.git` to clone this repo onto your local machine
- Navigate to the repo directory by running `cd ah-rambo-backend/`
- Ensure your working directory is clean by running `git status` command
- Get a copy of the PR by typing `git fetch origin pull/19/head:ft-validate-user-signup-inputs-167190529`
- Now that you have a copy of the branch, switch to it using `git checkout ft-validate-user-signup-inputs-167190529`
- Your directory will now be an exact copy of the PR
- Run `npm test`
- All tests should pass

#### What are the relevant pivotal tracker stories?
- [#167190529](https://www.pivotaltracker.com/story/show/167190529)

#### Any background context you want to add (Operations Impact)?
- Run `npm install` to get all dependencies onto your local machine

#### What I have learned working on this feature:
- As it was my first time unit testing a middleware, I have learned how to do that now.

#### Screenshots:
Screenshot 2019-08-01 at 9 37 03 AM
Screenshot 2019-08-01 at 9 38 37 AM

### CODE REVIEW

File: server/helpers/serverResponse.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/**
```
Comment (dinobi) [inline server/helpers/serverResponse.js]: please remove this file

Comment (abdulfatai360) [reply server/helpers/serverResponse.js]: @dinobi, I have updated my PR, and I noticed that the file has been updated on my branch as well. Basically, my local branch is in sync with the remote develop, so this file is no longer a repeated file.

You would notice that this file is no longer in the `Files changed` section of this PR.

**Review by chialuka — COMMENTED** [review]
Great job @abdulfatai360. Please, could you confirm that your work will throw an error when an unallowed variable is passed into the request body?

**Review by Adekoreday — CHANGES_REQUESTED** [review]
Good job, setCustomMessage would have been better placed outside the userSignup.js 
so it can be reusable for other input validations such as User Login, etc

**Comment by abdulfatai360** [comment]
> Great job @abdulfatai360. Please, could you confirm that your work will throw an error when an unallowed variable is passed into the request body?

@LukasChiama Thank you for your feedback. If you could remember, the decision to prevent an unwanted field from getting pass our validation middleware was why we changed from using `express-validator` to `Joi` as validation library. So I took that into serious consideration when implementing the feature.

According to Joi documentation, the `allowUnknown` property is set to `false` by default.

Refer to the link and screenshot below for evidences:
https://github.com/hapijs/joi/blob/master/API.md#anyvalidatevalue-options
Screenshot 2019-08-01 at 2 32 20 PM

**Comment by abdulfatai360** [comment]
> Good job, `setCustomMessage` would have been better placed outside the userSignup.js
> so it can be reusable for other input validations such as User Login, etc

@Adekoreday Thanks for your feedback. What you pointed out will sure improve the code quality of our codebase.

I have implemented your feedback. And push the changes to the repo. Thanks, once again.

File: server/schemas/userSignup.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    .max(254)
+    .regex(/^[\w._]+@[\w]+[-.]?[\w]+\.[\w]+$/)
+    .error(setCustomMessage('email')),
+  password: Joi.string()
+    .required()
+    .min(8)
+    .max(254)
+    .regex(/\s/, { invert: true })
+    .error(setCustomMessage('password'))
+};
```
Comment (chialuka) [inline server/schemas/userSignup.js:35]: Please could you include the `confirmPassword` field here as well? It is required for a check in the controllers for checking that passwords match.

**Review by henryade — APPROVED** [review]

**Review by Adekoreday — APPROVED** [review]

File: server/schemas/userSignup.js (hunk 2 above)
Comment (dinobi) [reply server/schemas/userSignup.js:35]: @LukasChiama I don't think that's part of our User schema

**Review by dinobi — APPROVED** [review]

**Pull request merged**